File: 2023/Rust/day-01-b/src/main.rs
use std::fs;
fn main() {
    const NUMBERS: [(&str,&str);9] = [
        ("one", "1"),
        ("two", "2"),
        ("three", "3"),
        ("four", "4"),
        ("five", "5"),
        ("six", "6"),
        ("seven", "7"),
        ("eight", "8"),
        ("nine", "9"),
    ];
    let mut input = match fs::read_to_string("../../inputs/input-01.txt") {
        Ok(value) => value,
        Err(_) => String::from("error"),
    };

    for (string, digit) in NUMBERS.iter() {
        let s = format!("{}{}{}",string,digit,string); 
        input = input.replace(string, &s);
    }

    let mut counter = 0; 
    for (index, line) in input.lines().enumerate() {
        let value = get_value(line);
        println!("{}: {}", index, value);
        counter += value;
    }
    println!("{}", counter);
}
fn get_value(string: &str) -> u32 {
    let mut first_digit: u32 = 0;
    let mut last_digit: u32 = 0;
    for c in string.chars() {
        if c.is_digit(10) {
            first_digit = c.to_digit(10).unwrap();
            break;
        }
    }
    for c in string.chars().rev() {
        if c.is_digit(10) {
            last_digit = c.to_digit(10).unwrap();
            break;
        }
    }
    first_digit * 10 + last_digit
}

File: 2023/Haskell/day-01-b/src/Main.hs
import Data.List (tails, isPrefixOf)
calculateValue :: [Int] -> Int
calculateValue digits = 10 * head digits + last digits

patterns :: [(String, Int)]
patterns = [
    ("1", 1), ("2", 2), ("3", 3), ("4", 4), ("5", 5), ("6", 6), ("7", 7), ("8", 8), ("9", 9),
    ("one", 1), ("two", 2), ("three", 3), ("four", 4), ("five", 5), ("six", 6), ("seven", 7), ("eight", 8), ("nine", 9)
    ]

solve :: [String] -> Int
solve = sum . map (calculateValue . numbers)
    where
    numbers :: String -> [Int] 
    numbers line = [n | string_tail <- tails line, (literal, n) <- patterns, literal `isPrefixOf` string_tail]


main :: IO ()
main = do
    let filepath :: String = "../../inputs/input-01.txt"
    input :: String <- readFile filepath
    let parsed :: [String] = lines input 
    let output :: Int = solve parsed
    print output
